node/impl/full/eth.go

```diff
@@ -1039,6 +1041,163 @@ func (a *EthModule) EthTraceTransaction(ctx context.Context, txHash string) ([]*
+	fromBlock, err := a.EthBlockNumberFromString(ctx, filter.FromBlock)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot parse fromBlock: %w", err)
+	}
+
+	toBlock, err := a.EthBlockNumberFromString(ctx, filter.ToBlock)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot parse toBlock: %w", err)
+	}
+
+	var results []*ethtypes.EthTraceFilterResult
+
+	if filter.Count != nil {
+		// If filter.Count is specified and it is 0, return an empty result set immediately.
+		if *filter.Count == 0 {
+			return []*ethtypes.EthTraceFilterResult{}, nil
+		}
+
+		// If filter.Count is specified and is greater than the EthTraceFilterMaxResults config return error
+		if *filter.Count > a.EthTraceFilterMaxResults {
+			return nil, xerrors.Errorf("invalid response count")
+		}
+	}
+
+	traceCounter := ethtypes.EthUint64(0)
+	for blkNum := fromBlock; blkNum <= toBlock; blkNum++ {
+		blockTraces, err := a.EthTraceBlock(ctx, strconv.FormatUint(uint64(blkNum), 10))
+		if err != nil {
+			return nil, xerrors.Errorf("cannot get trace for block %d: %w", blkNum, err)
+		}
+
+		for _, blockTrace := range blockTraces {
+			match, err := matchFilterCriteria(blockTrace, filter, filter.FromAddress, filter.ToAddress)
+			if err != nil {
+				return nil, xerrors.Errorf("cannot match filter for block %d: %w", blkNum, err)
+			}
+			if match {
+				traceCounter++
+				if filter.After != nil && traceCounter <= *filter.After {
+					continue
+				}
+
+				txTrace := ethtypes.EthTraceFilterResult{
+					EthTrace:            blockTrace.EthTrace,
+					BlockHash:           blockTrace.BlockHash,
+					BlockNumber:         blockTrace.BlockNumber,
+					TransactionHash:     blockTrace.TransactionHash,
+					TransactionPosition: blockTrace.TransactionPosition,
+				}
+				results = append(results, &txTrace)
+
+				// If Count is specified, limit the results
+				if filter.Count != nil && ethtypes.EthUint64(len(results)) >= *filter.Count {
+					return results, nil
+				}
+			}
+		}
+	}
+
+	return results, nil
+}
+
+// matchFilterCriteria checks if a trace matches the filter criteria.
+func matchFilterCriteria(trace *ethtypes.EthTraceBlock, filter ethtypes.EthTraceFilterCriteria, fromDecodedAddresses []ethtypes.EthAddress, toDecodedAddresses []ethtypes.EthAddress) (bool, error) {
+
+	var traceTo ethtypes.EthAddress
+	var traceFrom ethtypes.EthAddress
+
+	switch trace.Type {
+	case "call":
+		action, ok := trace.Action.(*ethtypes.EthCallTraceAction)
+		if !ok {
+			return false, xerrors.Errorf("invalid call trace action")
+		}
+		traceTo = action.To
+		traceFrom = action.From
+	case "create":
+		result, okResult := trace.Result.(*ethtypes.EthCreateTraceResult)
+		if !okResult {
+			return false, xerrors.Errorf("invalid create trace result")
+		}
+
+		action, okAction := trace.Action.(*ethtypes.EthCreateTraceAction)
+		if !okAction {
+			return false, xerrors.Errorf("invalid create trace action")
+		}
+
+		traceTo = *result.Address
```

Comment:
nit: a sanity nil pointer check would be nice.

Reply:
3b30a5553 addressed